src/lib/operators/sort.cpp

```diff
@@ -33,19 +20,78 @@ const std::string& Sort::name() const {
+  return std::make_shared<Sort>(copied_input_left, _sort_definitions, _output_chunk_size);
 }
 
 void Sort::_on_set_parameters(const std::unordered_map<ParameterID, AllTypeVariant>& parameters) {}
 
 std::shared_ptr<const Table> Sort::_on_execute() {
-  _impl = make_unique_by_data_type<AbstractReadOnlyOperatorImpl, SortImpl>(
-      input_table_left()->column_data_type(_column_id), input_table_left(), _column_id, _order_by_mode,
-      _output_chunk_size);
-  return _impl->_on_execute();
+  _sort_definition_data_types.resize(_sort_definitions.size());
+  for (auto column_id = ColumnID{0}; column_id < _sort_definitions.size(); ++column_id) {
+    auto sort_definition = _sort_definitions[column_id];
+    _sort_definition_data_types[column_id] = input_table_left()->column_data_type(sort_definition.column);
+  }
+
+  std::shared_ptr<const Table> output = input_table_left();
+  std::shared_ptr<PosList> previously_sorted_pos_list = std::shared_ptr<PosList>(nullptr);
+
+  _validate_sort_definitions();
+
+  for (auto column_id = _sort_definitions.size(); column_id-- != 0;) {
+    const bool is_last_sorting_run = (column_id == 0);
+
+    auto sort_definition = _sort_definitions[column_id];
+    DataType data_type = _sort_definition_data_types[column_id];
+
+    resolve_data_type(data_type, [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+
+      auto sort_single_col_impl = SortImpl<ColumnDataType>(input_table_left(), sort_definition.column,
+                                                           sort_definition.order_by_mode, _output_chunk_size);
+      const auto row_id_value_vector = sort_single_col_impl.sort_one_column(previously_sorted_pos_list);
+
+      // 3. Materialization of the result: We take the sorted ValueRowID Vector, create chunks fill them until they are
+      // full and create the next one. Each chunk is filled row by row.
+      if (is_last_sorting_run) {
+        auto materialization = std::make_shared<SortImplMaterializeOutput<ColumnDataType>>(
+            input_table_left(), row_id_value_vector, _output_chunk_size);
+
+        std::shared_ptr<Table> output_mutable = materialization->execute();
+
+        // Set the ordered_by attribute of the output's chunks to the column the chunk was sorted by last (because this
+        // is the column that the chunk is ordered by in any case; whereas it is not necessarily sorted by the other
+        // columns -- on their own -- as well).
+        const auto chunk_count = output_mutable->chunk_count();
+        for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+          output_mutable->get_chunk(chunk_id)->set_ordered_by(
+              std::make_pair(sort_definition.column, sort_definition.order_by_mode));
+        }
+        output = output_mutable;
+      } else {
+        previously_sorted_pos_list.reset();
+        previously_sorted_pos_list = std::make_shared<PosList>();
```

Comment:
Regarding std::shared_ptr::reset:
`For signature (1) the object becomes empty (as if default-constructed).`

So I guess, one of these lines (71 or 72) is enough

Source: [CPP-reference](https://www.cplusplus.com/reference/memory/shared_ptr/reset/)